src/graph/network/socket_communicator.cc

```diff
@@ -57,101 +86,201 @@ bool SocketSender::Connect() {
+  // Check address format
+  if (substring[0] != "socket:" || substring.size() != 2) {
+    LOG(FATAL) << "Incorrect address format:" << addr
+               << " Please provide right address format, "
+               << "e.g, 'socket://127.0.0.1:50051'. ";
+  }
+  // Get IP and port
+  SplitStringUsing(substring[1], ":", &ip_and_port);
+  if (ip_and_port.size() != 2) {
+    LOG(FATAL) << "Incorrect address format:" << addr
+               << " Please provide right address format, "
+               << "e.g, 'socket://127.0.0.1:50051'. ";
+  }
+  std::string ip = ip_and_port[0];
+  int port = stoi(ip_and_port[1]);
+  // Initialize message queue for each connection
   num_sender_ = num_sender;
-  queue_size_ = queue_size;
-  queue_ = new MessageQueue(queue_size_, num_sender_);
-  // Initialize socket, and socket_[0] is server socket
-  socket_.resize(num_sender_+1);
-  thread_.resize(num_sender_);
-  socket_[0] = new TCPSocket();
-  TCPSocket* server = socket_[0];
-  server->SetTimeout(kTimeOut * 60 * 1000);  // millsec
+  for (int i = 0; i < num_sender_; ++i) {
+    msg_queue_[i] = new MessageQueue(queue_size_);
+  }
+  // Initialize socket and socket-thread
+  server_socket_ = new TCPSocket();
+  server_socket_->SetTimeout(kTimeOut * 60 * 1000);  // millsec
   // Bind socket
-  if (server->Bind(ip, port) == false) {
+  if (server_socket_->Bind(ip.c_str(), port) == false) {
     LOG(FATAL) << "Cannot bind to " << ip << ":" << port;
     return false;
   }
   LOG(INFO) << "Bind to " << ip << ":" << port;
   // Listen
-  if (server->Listen(kMaxConnection) == false) {
+  if (server_socket_->Listen(kMaxConnection) == false) {
     LOG(FATAL) << "Cannot listen on " << ip << ":" << port;
     return false;
   }
   LOG(INFO) << "Listen on " << ip << ":" << port << ", wait sender connect ...";
   // Accept all sender sockets
   std::string accept_ip;
   int accept_port;
-  for (int i = 1; i <= num_sender_; ++i) {
-    socket_[i] = new TCPSocket();
-    if (server->Accept(socket_[i], &accept_ip, &accept_port) == false) {
+  for (int i = 0; i < num_sender_; ++i) {
+    sockets_[i] = new TCPSocket();
+    if (server_socket_->Accept(sockets_[i], &accept_ip, &accept_port) == false) {
       LOG(FATAL) << "Error on accept socket.";
       return false;
     }
     // create new thread for each socket
-    thread_[i-1] = new std::thread(MsgHandler, socket_[i], queue_, i-1);
+    threads_[i] = new std::thread(RecvLoop, sockets_[i], msg_queue_[i]);
     LOG(INFO) << "Accept new sender: " << accept_ip << ":" << accept_port;
   }
 
   return true;
 }
 
-void SocketReceiver::MsgHandler(TCPSocket* socket, MessageQueue* queue, int id) {
-  char* buffer = new char[kMaxBufferSize];
+char* SocketReceiver::Recv(int64_t* size, int* send_id) {
+  // Get the top message
+  for (;;) {  // loop until get a message
+    for (auto& mq : msg_queue_) {
+      if (mq.second->Empty() == false) {
+        *send_id = mq.first;
+        return msg_queue_[*send_id]->Remove(size);
```

Comment:
That is a question. In your case, the Remove() will be blocked until its msg_queue is not empty. For our case, we only have one thread to invoke the recv() API, so that is not a problem. For multi-threading, do you have any suggestion to make it efficient?  It is too expensive if we use a global mutex here,

Reply:
Well, technically we are using global mutex already when entering `MessageQueue::Remove`. The problem here could be solved by letting `Remove` returns size 0 if the queue is empty:
```c++
for (;;) {
  for (auto& mq : msg_queue_) {
    *send_id = mq.first;
    const int64_t size = mq.second->Remove(msg);
    if (size < 0) {
      LOG(..) << ..;
      return -1;
    }
    if (size > 0) {
      return size;
    }
  }
}
```
Also, I feel quite redundant to return the data size when the `Message` structure already contains a size. Consider return some status code?

Reply:
For example, SocketComminactor is thread-safe but it's not in MPI implementation.